# Lab 1: Arrays and DataFrames

Welcome to Lab 1! This week, we'll learn about arrays, which allow us to store sequences of data, and DataFrames, which let us work with multiple arrays of data about the same things. You should complete this entire lab so that all tests pass.

**Please do not use for-loops for any questions in this lab.** If you don't know what a for-loop is, don't worry – we haven't covered them yet. But if you do know what they are and are wondering why it's not OK to use them, it is because loops in Python are slow, and looping over arrays and DataFrames should usually be avoided.

First, set up the imports we'll need by running the cell below.

In [1]:
in_colab = "google.colab" in str(get_ipython())
if in_colab:
    !pip install otter-grader==6.1.6

from pathlib import Path
import numpy as np
np.set_printoptions(legacy='1.21')
import pandas as pd

path = 'labs/lab01/build/student'
if in_colab:
    import shutil
    import zipfile

    assets = ['data', 'tests', 'images']
    missing_assets = [asset for asset in assets if not Path(asset).exists()]
    if missing_assets:
        !wget -q -O /content/course.zip https://github.com/dsc-courses/cosmos-ml-cluster-2026/archive/refs/heads/main.zip
        with zipfile.ZipFile('/content/course.zip') as course_zip:
            for asset in missing_assets:
                archive_prefix = f"cosmos-ml-cluster-2026-main/{path}/{asset}/"
                members = [name for name in course_zip.namelist() if name.startswith(archive_prefix)]
                if members:
                    course_zip.extractall('/content/course-assets', members)
                    source_path = Path('/content/course-assets') / archive_prefix
                    if source_path.exists():
                        shutil.copytree(source_path, asset, dirs_exist_ok=True)

import otter
grader = otter.Notebook()

# 1. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That is, if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

```py
0.18 * billions_of_numbers
```

evaluates to a new array of numbers that's the result of multiplying each number in `billions_of_numbers` by 0.18 (18%).  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**, like a column in a spreadsheet (think Google Sheets or Microsoft Excel). 

<img src="data/sheet_array.png" width=600>

## 1.1. Making arrays
You can type in the data that goes in an array yourself, but that's not typically how we'll create arrays. Normally, we create arrays by loading them from an external source, like a data file.

First, though, let's learn how to do it the hard way. To begin, we can make a **list** of numbers by putting them within square brackets and separating them by commas:

In [2]:
my_list = [14, -2.26, 0.15]
my_list

[14, -2.26, 0.15]

Just like `int`, `float`, and `str`, the `list` is a data type provided by Python. Lists are very flexible and easy to work with, but they are *slowwww* 🐢.

As data scientists, we'll often be working with millions or even billions of numbers. For this, we need something faster than a `list`. Instead of lists, we will use *arrays*. 

Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie" or, if you prefer to pronounce things incorrectly, "NUM-pee").  The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Data scientists, as well as engineers and scientists of all kinds, use `numpy` frequently, and you'll see quite a bit of it if you're a data science major.

In [3]:
import numpy as np

Now, to create an array, call the function `np.array` with a list of numbers.  Run this cell to see an example:

In [4]:
np.array([14, -2.26, 0.15])

array([14.  , -2.26,  0.15])

Note that you need the square-brackets here. If you were to try running the following code, Python would yell at you because you forgot them:

```py
np.array(14, -2.26, 0.15)
```

<img src='data/brackets.png' width=400>

Arrays themselves are also values, just like numbers and strings.  That means you can assign them names or use them as arguments to functions.

**Question 1.1.1.** Make an array containing the numbers 2, 4, and 6, in that order.  Name it `even_numbers`.

In [5]:
even_numbers = np.array([2, 4, 6]) # SOLUTION
even_numbers

array([2, 4, 6])

In [6]:
isinstance(even_numbers, np.ndarray)

True

In [7]:
len(even_numbers) == 3

True

In [8]:
all(even_numbers == np.arange(2, 8, 2))

True

**Question 1.1.2.** Make an array containing the numbers 0, -1, 1, $\pi$, and $e$, in that order.  Name it `odd_numbers`.

**_Hint:_**  $\pi$ and $e$ are available from the `np` module, which has already been imported. Just as you used `math.pi` to get $\pi$ in the last lab, you can use `np.pi` to get $\pi$ as well. **Do not** import the `math` module.

In [9]:
odd_numbers = np.array([0, -1, 1, np.pi, np.e]) # SOLUTION
odd_numbers

array([ 0.        , -1.        ,  1.        ,  3.14159265,  2.71828183])

In [10]:
isinstance(odd_numbers, np.ndarray)

True

In [11]:
len(odd_numbers) == 5

True

In [12]:
all(odd_numbers == np.array([0, -1, 1, np.pi, np.e]))

True

**Question 1.1.3.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you print `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely odd way of saying that the things in the array are strings. In case you're interested, the `U` means that this string is encoded in [unicode](https://en.wikipedia.org/wiki/Unicode), and the `<5` means all strings in the array are 5 characters long or less.

In [13]:
hello_world_components = np.array(['Hello', ',', ' ', 'world', '!']) # SOLUTION
hello_world_components

array(['Hello', ',', ' ', 'world', '!'], dtype='<U5')

In [14]:
hello_world_components[0] == 'Hello'

True

In [15]:
hello_world_components[2] == ' '

True

In [16]:
hello_world_components[4] == '!'

True

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The expression `np.arange(start, stop, space)` produces an array with all the numbers starting at `start` and counting up by `space`, stopping **before** `stop` is reached.

For example, the value of `np.arange(1, 8, 2)` is an array with elements 1, 3, 5, and 7 – it starts at 1 and counts up by 2, then stops before 8.  In other words, it makes the same array as `np.array([1, 3, 5, 7])`.

`np.arange(4, 9, 1)` is an array with elements 4, 5, 6, 7, and 8.  (It doesn't contain 9 because `np.arange` stops *before* the stop value is reached.)

**Question 1.1.4.** Use `np.arange` to create an array with all the multiples of 99 from 0 up to (**and including**) 9999.  (Its elements should be 0, 99, 198, 297, and so on.)

In [17]:
multiples_of_99 = np.arange(0, 10_000, 99) # SOLUTION
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [18]:
type(multiples_of_99) == np.ndarray

True

In [19]:
len(multiples_of_99) == 102

True

In [20]:
all(multiples_of_99 == np.arange(0, 9999+99, 99))

True

##### Temperature readings 🌡️
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the San Diego International Airport weather station for the month of September 2023. To analyze the data, we want to know when each reading was taken, but we find that the data doesn't include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of September 2023 (midnight on September 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 1.1.5.** Create an array of the *time, in seconds, since the start of the month* at which each hourly reading was taken.  Name it `collection_times`.

**_Hints:_** 
- There are 30 days in September, which is equivalent to ($30 \times 24$) hours or ($30 \times 24 \times 60 \times 60$) seconds.  
- The `len` function works on arrays, too.  If your `collection_times` isn't passing the tests, check its length and make sure it has $30 \times 24$ elements, since readings are taken hourly for 30 days.

In [21]:
collection_times = np.arange(0, 30 * 24 * 60 * 60, 60 * 60) # SOLUTION
collection_times

array([      0,    3600,    7200,   10800,   14400,   18000,   21600,
         25200,   28800,   32400,   36000,   39600,   43200,   46800,
         50400,   54000,   57600,   61200,   64800,   68400,   72000,
         75600,   79200,   82800,   86400,   90000,   93600,   97200,
        100800,  104400,  108000,  111600,  115200,  118800,  122400,
        126000,  129600,  133200,  136800,  140400,  144000,  147600,
        151200,  154800,  158400,  162000,  165600,  169200,  172800,
        176400,  180000,  183600,  187200,  190800,  194400,  198000,
        201600,  205200,  208800,  212400,  216000,  219600,  223200,
        226800,  230400,  234000,  237600,  241200,  244800,  248400,
        252000,  255600,  259200,  262800,  266400,  270000,  273600,
        277200,  280800,  284400,  288000,  291600,  295200,  298800,
        302400,  306000,  309600,  313200,  316800,  320400,  324000,
        327600,  331200,  334800,  338400,  342000,  345600,  349200,
        352800,  356

In [22]:
type(collection_times) == np.ndarray

True

In [23]:
len(collection_times) == 720

True

In [24]:
np.array_equal(collection_times, np.arange(0, 30 * 24 * 60 * 60, 60 * 60))

True

## 1.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population` that includes estimated world populations in every year from **1950** to **2023**.  (The estimates come from the [International Database](https://www.census.gov/data-tools/demo/idb/#/table?menu=tableViz&quickReports=CUSTOM&CUSTOM_COLS=POP,TFR,CBR,E0,IMR,CDR,NMR&CCODE=**&show_countries=n&CCODE_SINGLE=**&TABLE_RANGE=1950,2023&TABLE_YEARS=1950,1951,1952,1953,1954,1955,1956,1957,1958,1959,1960,1961,1962,1963,1964,1965,1966,1967,1968,1969,1970,1971,1972,1973,1974,1975,1976,1977,1978,1979,1980,1981,1982,1983,1984,1985,1986,1987,1988,1989,1990,1991,1992,1993,1994,1995,1996,1997,1998,1999,2000,2001,2002,2003,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023&TABLE_USE_RANGE=Y&TABLE_USE_YEARS=N&TABLE_STEP=1), maintained by the US Census Bureau.)

Rather than type in the data manually, we've loaded them from a file called `world_population_2023.csv`.  You'll learn how to read in data from files very soon.

In [25]:
# Don't worry too much about what goes on in this cell.
population = pd.read_csv("data/world_population_2023.csv").get("Population").values
population

array([2558023014, 2595838116, 2637936830, 2683288876, 2731379529,
       2782973591, 2836130358, 2892165757, 2948892474, 3001437149,
       3043723411, 3084753192, 3140982059, 3210747452, 3282150912,
       3351447909, 3421950085, 3491918938, 3563999493, 3638882226,
       3714262611, 3791705363, 3867790515, 3943299941, 4018289723,
       4089897507, 4160016430, 4230925634, 4301796324, 4375462721,
       4446018801, 4527470698, 4610284677, 4694209205, 4775889225,
       4860691329, 4947831057, 5037623680, 5128370129, 5218863705,
       5311093985, 5398248391, 5484859952, 5568560590, 5650427498,
       5733476988, 5815625725, 5896157461, 5975537190, 6054374592,
       6133007860, 6211824800, 6290903765, 6369899427, 6449064951,
       6528030077, 6608491622, 6690682690, 6774892222, 6859055794,
       6942148774, 7024885273, 7108237070, 7192338928, 7276120452,
       7358968084, 7441688089, 7523957647, 7605030254, 7685586259,
       7764965463, 7837646014, 7906702795, 7982019198])

Here's how we get the first element of `population`, which is the world population in the first year in the dataset, 1950.

In [26]:
population[0]

2558023014

Notice that we use square brackets here. The square brackets signal that we are *accessing* an element of the array. Square brackets in Python are kind of like subscripts in math.

The value of that expression is the number 2558023014 (around 2.5 billion), because that's the first thing in the array `population`.

Notice that we wrote `population[0]`, not `population[1]`, to get the first element.  This is a weird convention in programming. 0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population`.  Read and run each cell.

In [27]:
# The third element in the array is the population in 1952.
population_1952 = population[2]
population_1952

2637936830

In [28]:
# The thirteenth element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population[12]
population_1962

3140982059

In [29]:
# The 74th element in the array is the population in 2023.
population_2023 = population[73]
population_2023

7982019198

In [30]:
# The array has only 74 elements, so this doesn't work.
# (There's no element with 74 other elements before it.)

# population_2024 = population[74]
# population_2024

# 🚨 After running this cell, please place a # before each line above to make sure that it doesn't run again.

**Question 1.2.1.** Set `population_1998` to the world population in 1998 by getting the appropriate element from `population`.

In [31]:
population_1998 = population[48] # SOLUTION
population_1998

5975537190

In [32]:
bool(population_1998 == 5975537190)

True

## 1.3. Performing an operation on every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to access and work with single elements.

##### Logarithms
Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

The logarithm function is one way of measuring how big a number is. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from NumPy on each element of the `population` array:

In [33]:
population_1950_magnitude = np.log10(population[0])
population_1951_magnitude = np.log10(population[1])
population_1952_magnitude = np.log10(population[2])
population_1953_magnitude = np.log10(population[3])

# ... and so on!

But this is tedious and repetitive. There must be a better way!

It turns out that NumPy's `log10` is pretty powerful. Not only can it take in a single number (like `population[0]`) as input and return the logarithm of a single number, but it can **also** take in an entire array of numbers and return the logarithm of each element in that array!

If you give NumPy's `log10` an array as input, it will return an array of the same length, where the first element of the result is the logarithm of the first element of the input, the second element of the result is the logarithm of the second element of the input, and so on.

<img src="data/array_logarithm.jpg">

This is called *elementwise* application of the function, since it operates separately on each element of the array it's called on.  

**Question 1.3.1.** Use NumPy's `log10` function to compute the logarithms of the world population in every year.  Give the result (an array of 73 numbers) the name `population_magnitudes`.  Your code should be very short.

In [34]:
population_magnitudes = np.log10(population) # SOLUTION
population_magnitudes

array([9.40790445, 9.41427761, 9.42126439, 9.42866743, 9.43638205,
       9.44450909, 9.45272619, 9.46122318, 9.46965894, 9.47732925,
       9.48340518, 9.48922042, 9.49706546, 9.50660615, 9.51615855,
       9.52523247, 9.53427367, 9.54306415, 9.55193763, 9.560968  ,
       9.56987261, 9.57883458, 9.58746294, 9.59585981, 9.60404125,
       9.61171242, 9.61909505, 9.62643539, 9.63364984, 9.64102399,
       9.64797129, 9.65585565, 9.66372774, 9.67156244, 9.67905424,
       9.68669804, 9.69441486, 9.70222572, 9.70997936, 9.71757596,
       9.72518399, 9.73225286, 9.73916554, 9.74574295, 9.75208131,
       9.75841807, 9.76459645, 9.77056907, 9.77637695, 9.78206929,
       9.78767352, 9.7932192 , 9.79871304, 9.80413258, 9.80949675,
       9.81478215, 9.82010234, 9.82547043, 9.83090239, 9.83626434,
       9.84149392, 9.84663924, 9.8517619 , 9.85687014, 9.86189988,
       9.86681692, 9.87167146, 9.87644634, 9.88110095, 9.885677  ,
       9.89013953, 9.89418564, 9.89799541, 9.90211277])

In [35]:
type(population_magnitudes) == np.ndarray

True

In [36]:
len(population_magnitudes) == 74

True

In [37]:
bool(sum(abs(population_magnitudes - np.log10(population))) < 1e-6)

True

##### Arithmetic
Arithmetic also works elementwise on arrays.  For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [38]:
population_in_billions = population / 1000000000
population_in_billions

array([2.55802301, 2.59583812, 2.63793683, 2.68328888, 2.73137953,
       2.78297359, 2.83613036, 2.89216576, 2.94889247, 3.00143715,
       3.04372341, 3.08475319, 3.14098206, 3.21074745, 3.28215091,
       3.35144791, 3.42195009, 3.49191894, 3.56399949, 3.63888223,
       3.71426261, 3.79170536, 3.86779051, 3.94329994, 4.01828972,
       4.08989751, 4.16001643, 4.23092563, 4.30179632, 4.37546272,
       4.4460188 , 4.5274707 , 4.61028468, 4.6942092 , 4.77588923,
       4.86069133, 4.94783106, 5.03762368, 5.12837013, 5.21886371,
       5.31109399, 5.39824839, 5.48485995, 5.56856059, 5.6504275 ,
       5.73347699, 5.81562573, 5.89615746, 5.97553719, 6.05437459,
       6.13300786, 6.2118248 , 6.29090377, 6.36989943, 6.44906495,
       6.52803008, 6.60849162, 6.69068269, 6.77489222, 6.85905579,
       6.94214877, 7.02488527, 7.10823707, 7.19233893, 7.27612045,
       7.35896808, 7.44168809, 7.52395765, 7.60503025, 7.68558626,
       7.76496546, 7.83764601, 7.9067028 , 7.9820192 ])

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a twenty percent tip on several restaurant bills at once:

In [39]:
restaurant_bills = np.array([20.12, 39.90, 31.01])
print("Restaurant bills:\t", restaurant_bills)
tips = 0.2 * restaurant_bills 
print("Tips:\t\t\t", tips)

Restaurant bills:	 [20.12 39.9  31.01]
Tips:			 [4.024 7.98  6.202]


<img src="data/array_multiplication.jpg">

**Question 1.3.2.** Suppose the total charge at a restaurant is the original bill plus the tip (20%).  That means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills` and give the resulting array the name `total_charges`.

In [40]:
total_charges = restaurant_bills * 1.2 # SOLUTION
total_charges

array([24.144, 47.88 , 37.212])

In [41]:
type(total_charges) == np.ndarray

True

In [42]:
bool(sum(abs(total_charges - np.array([24.144, 47.88, 37.212]))) < 1e-6)

True

Let's read in some data to use in the next question.

In [43]:
more_restaurant_bills = pd.read_csv("data/more_restaurant_bills.csv").get("Bill").values

**Question 1.3.3.** The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one, assuming again a twenty percent tip, and give the resulting array the name `more_total_charges`.

In [44]:
more_total_charges = more_restaurant_bills * 1.2 # SOLUTION
more_total_charges

array([20.244, 20.892, 12.216, ..., 19.308, 18.336, 35.664])

In [45]:
type(more_total_charges) == np.ndarray

True

In [46]:
bool(sum(abs(more_total_charges - 1.2 * more_restaurant_bills)) < 1e-6)

True

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 1.3.4.** What was the sum of all the bills in `more_restaurant_bills`, **including tips**?

In [47]:
sum_of_bills = sum(more_total_charges) # SOLUTION
sum_of_bills

1795730.0640000193

In [48]:
bool(round(sum_of_bills, 2) == 1795730.06)

True

##### Powers of Two
The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or computers come in powers of 2, like 64 GB, 128 GB, or 256 GB.)

**Question 1.3.5.** Use `np.arange` and the exponentiation operator `**` to create an array containing the first 40 powers of 2, starting from $2^0=1$.

**_Hints:_**
- Did your kernel "die" when you ran your solution? There is a common incorrect response to this problem that tries to create an array with so many entries that Python gives up and crashes. If this happens to you, rethink your solution.
- To make it easier to see where you might be going, start by generating an array with the first 5 powers of 2, instead of the first 40. Once you've figure that out, then try all 40. At no point should you have to manually write `0, 1, 2, 3, 4, ...`; if you find yourself trying that, scroll up to earlier in the lab notebook.

In [49]:
powers_of_2 = 2 ** np.arange(40) # SOLUTION
powers_of_2

array([           1,            2,            4,            8,
                 16,           32,           64,          128,
                256,          512,         1024,         2048,
               4096,         8192,        16384,        32768,
              65536,       131072,       262144,       524288,
            1048576,      2097152,      4194304,      8388608,
           16777216,     33554432,     67108864,    134217728,
          268435456,    536870912,   1073741824,   2147483648,
         4294967296,   8589934592,  17179869184,  34359738368,
        68719476736, 137438953472, 274877906944, 549755813888])

In [50]:
len(powers_of_2) == 40

True

In [51]:
np.array_equal(powers_of_2, 2 ** np.arange(40))

True

# 2. DataFrames 

## 2.1. Introduction

For a collection of things in the world, an array is useful for describing a single attribute of each thing. For example, among the collection of US States, an array could describe the land area of each. Tables extend this idea by describing multiple attributes for each element of a collection. In a table of states, for example, we might keep track of land area, population, state capital, and the name of the governor. In other words, tables keep track of many entities (individuals, stored as rows), and for each entity, many attributes (features, stored as columns).

In the cell below we have two arrays. The first one contains the world population in each year (as estimated by the US Census Bureau), and the second contains the years themselves (in order, so the first elements in the population and the years arrays correspond).

In [52]:
population_amounts = pd.read_csv("data/world_population_2023.csv").get("Population").values
population_years = np.arange(1950, 2023 + 1)
print("Population column:", population_amounts)
print("Years column:", population_years)

Population column: [2558023014 2595838116 2637936830 2683288876 2731379529 2782973591
 2836130358 2892165757 2948892474 3001437149 3043723411 3084753192
 3140982059 3210747452 3282150912 3351447909 3421950085 3491918938
 3563999493 3638882226 3714262611 3791705363 3867790515 3943299941
 4018289723 4089897507 4160016430 4230925634 4301796324 4375462721
 4446018801 4527470698 4610284677 4694209205 4775889225 4860691329
 4947831057 5037623680 5128370129 5218863705 5311093985 5398248391
 5484859952 5568560590 5650427498 5733476988 5815625725 5896157461
 5975537190 6054374592 6133007860 6211824800 6290903765 6369899427
 6449064951 6528030077 6608491622 6690682690 6774892222 6859055794
 6942148774 7024885273 7108237070 7192338928 7276120452 7358968084
 7441688089 7523957647 7605030254 7685586259 7764965463 7837646014
 7906702795 7982019198]
Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 19

Suppose we want to answer this question:

> When did the world's population surpass 7 billion?

You could technically answer this question just by staring at the arrays, but it's a bit complicated, since you would have to count the position where the population first crossed 7 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a table.

Just as `numpy` provides arrays, a popular package called `pandas` provides **DataFrames**, which is `pandas`' name for **tables**. `pandas` is *the* tool for doing data science in Python.

You can import `pandas` using the following code:

In [53]:
import pandas as pd

We'll use the standard abbreviation `pd` whenever we work with pandas in this lab. If you continue doing data analysis in Python, you'll see `pandas` often.

The cell below:

- creates an empty DataFrame using the expression `pd.DataFrame()`,
- assigns two columns to the DataFrame by calling `assign`,
- assigns the resulting DataFrame to the name `population_df`, and finally
- displays `population_df` so that we can see the DataFrame we've made.

`"Population"` and `"Year"` are column labels that we have chosen. We could have chosen anything, but it's a good idea to choose names that are descriptive and not too long.

In [54]:
population_df = pd.DataFrame().assign(
    Population=population_amounts,
    Year=population_years
)
population_df

,Population,Year
0,2558023014,1950
1,2595838116,1951
2,2637936830,1952
3,2683288876,1953
4,2731379529,1954
...,...,...
69,7685586259,2019
70,7764965463,2020
71,7837646014,2021
72,7906702795,2022


Now the data are all together in a single DataFrame! It's much easier to parse this data. If you need to know what the population was in 2011, for example, you can tell from a single glance. We'll revisit this DataFrame later, but first we'll build some skills practicing with a new DataFrame of top movies.

**Question 2.1.1.** In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a DataFrame that has two columns called `"Rating"` and `"Name"`, which hold `top_10_movie_ratings` and `top_10_movie_names` respectively.

In [55]:
top_10_movie_ratings = np.array([9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8])
top_10_movie_names = np.array([
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'
])

In [56]:
top_10_movies = pd.DataFrame().assign(Name=top_10_movie_names, Rating=top_10_movie_ratings) # SOLUTION
top_10_movies

,Name,Rating
0,The Shawshank Redemption (1994),9.2
1,The Godfather (1972),9.2
2,The Godfather: Part II (1974),9.0
3,Pulp Fiction (1994),8.9
4,Schindler's List (1993),8.9
5,The Lord of the Rings: The Return of the King ...,8.9
6,12 Angry Men (1957),8.9
7,The Dark Knight (2008),8.9
8,"Il buono, il brutto, il cattivo (1966)",8.9
9,The Lord of the Rings: The Fellowship of the R...,8.8


In [57]:
isinstance(top_10_movies, pd.DataFrame)

True

In [58]:
top_10_movies.shape == (10, 2)

True

In [59]:
set(top_10_movies.columns) == {'Name', 'Rating'}

True

Suppose you want to add your own ratings to this DataFrame. The cell below contains your ranking of each movie:

In [60]:
my_ranking = [8, 2, 1, 9, 7, 10, 6, 4, 3, 5]

**Question 2.1.2** You can use the `assign` method to add a column to an already-existing DataFrame, too. Create a new DataFrame called `with_ranking` by adding a column named `"Ranking"` to the DataFrame in `top_10_movies`.

In [61]:
with_ranking = top_10_movies.assign(Ranking=my_ranking) # SOLUTION
with_ranking

,Name,Rating,Ranking
0,The Shawshank Redemption (1994),9.2,8
1,The Godfather (1972),9.2,2
2,The Godfather: Part II (1974),9.0,1
3,Pulp Fiction (1994),8.9,9
4,Schindler's List (1993),8.9,7
5,The Lord of the Rings: The Return of the King ...,8.9,10
6,12 Angry Men (1957),8.9,6
7,The Dark Knight (2008),8.9,4
8,"Il buono, il brutto, il cattivo (1966)",8.9,3
9,The Lord of the Rings: The Fellowship of the R...,8.8,5


In [62]:
isinstance(with_ranking, pd.DataFrame)

True

In [63]:
with_ranking.shape == (10, 3)

True

In [64]:
set(with_ranking.columns) == {'Name', 'Rating', 'Ranking'}

True

## 2.2. Indexes

You may have noticed that the DataFrame of populations contains what looks like an extra, unlabeled column on the left with the numbers 0 through 73. **This is not a column, it's what we call an *index***. The index contains the row labels. Whereas the columns of this DataFrame are labeled `"Population"` and `"Year"`, the rows are labeled 0, 1, ..., 73.

By default, pandas uses row labels numbered 0, 1, 2, and so on. In this case, it makes more sense to use the year as a row's label. We can do this by setting the `"Year"` column as the index:

In [65]:
population_by_year = population_df.set_index('Year')
population_by_year

,Population
Year,
1950,2558023014
1951,2595838116
1952,2637936830
1953,2683288876
1954,2731379529
...,...
2019,7685586259
2020,7764965463
2021,7837646014


As we'll see, this does more than make the DataFrame look nicer – it is very useful, too. Let's perform this same process on the `top_10_movies` DataFrame we were just working with.

**Question 2.2.1** Create a new DataFrame named `top_10_movies_by_name` by taking the DataFrame you made above, `top_10_movies`, and setting the index to be the `'Name'` column.

In [66]:
top_10_movies_by_name = top_10_movies.set_index('Name') # SOLUTION
top_10_movies_by_name

,Rating
Name,
The Shawshank Redemption (1994),9.2
The Godfather (1972),9.2
The Godfather: Part II (1974),9.0
Pulp Fiction (1994),8.9
Schindler's List (1993),8.9
The Lord of the Rings: The Return of the King (2003),8.9
12 Angry Men (1957),8.9
The Dark Knight (2008),8.9
"Il buono, il brutto, il cattivo (1966)",8.9


In [67]:
top_10_movies_by_name.index[0] == 'The Shawshank Redemption (1994)'

True

You can get an array of row names using `.index`. For instance, the array of row names of the `population_by_year` DataFrame is:

In [68]:
population_by_year.index

RangeIndex(start=1950, stop=2024, step=1, name='Year')

**Question 2.2.2** Using code, assign to `tenth_movie` the name of the tenth movie in `top_10_movies_by_name`.

**_Hint:_**  Remember that the index is an array, and we use square brackets to access elements of an array.

In [69]:
tenth_movie = top_10_movies_by_name.index[9] # SOLUTION
tenth_movie

'The Lord of the Rings: The Fellowship of the Ring (2001)'

In [70]:
tenth_movie == 'The Lord of the Rings: The Fellowship of the Ring (2001)'

True

## 2.3 Reading a DataFrame from a file
In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we can use functions provided by pandas to read in data from external files.

The function `pd.read_csv` takes one argument, a path to a data file (a string), and returns a DataFrame.  There are many formats for data files, but CSV ("comma-separated values") is the most common. 

**Question 2.3.1.** The file `data/imdb.csv` contains information about the 250 highest-rated movies on IMDb.  Load it as a DataFrame called `imdb`.

In [71]:
imdb = pd.read_csv('data/imdb.csv') # SOLUTION
imdb

,Votes,Rating,Title,Year,Decade
0,88355,8.4,M,1931,1930
1,132823,8.3,Singin' in the Rain,1952,1950
2,74178,8.3,All About Eve,1950,1950
3,635139,8.6,Léon,1994,1990
4,145514,8.2,The Elephant Man,1980,1980
...,...,...,...,...,...
245,1078416,8.7,Forrest Gump,1994,1990
246,31003,8.1,Le salaire de la peur,1953,1950
247,167076,8.2,3 Idiots,2009,2000
248,91689,8.1,Network,1976,1970


In [72]:
isinstance(imdb, pd.DataFrame)

True

In [73]:
imdb.shape == (250, 5)

True

Notice the dots in the middle of the DataFrame. This means that a lot of the rows have been omitted. This DataFrame is big enough that only a few of its rows are displayed, but the others are still there.  There are 250 movies total.

Where did `imdb.csv` come from? Take a look at [this lab's folder](./). If you go into the `data/` directory, you should see a file called `imdb.csv`.

Open up the `imdb.csv` file in that folder and look at the format. What do you notice? The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

**Question 2.3.2.** This is a data set of movies, so it makes sense to use the movie title as the row label. Create a new DataFrame called `imdb_by_name` which uses the movie title as the index.

In [74]:
imdb_by_name = imdb.set_index('Title') # SOLUTION
imdb_by_name

,Votes,Rating,Year,Decade
Title,,,,
M,88355,8.4,1931,1930
Singin' in the Rain,132823,8.3,1952,1950
All About Eve,74178,8.3,1950,1950
Léon,635139,8.6,1994,1990
The Elephant Man,145514,8.2,1980,1980
...,...,...,...,...
Forrest Gump,1078416,8.7,1994,1990
Le salaire de la peur,31003,8.1,1953,1950
3 Idiots,167076,8.2,2009,2000


In [75]:
isinstance(imdb_by_name, pd.DataFrame)

True

In [76]:
imdb_by_name.shape == (250, 4)

True

In [77]:
imdb_by_name.index[0] == 'M'

True

## 2.4. Series



Suppose we're interested primarily in movie ratings. To extract just this column from the DataFrame, we use the `.get` method:

In [78]:
ratings = imdb_by_name.get('Rating')
ratings

Title
M                                        8.4
Singin' in the Rain                      8.3
All About Eve                            8.3
Léon                                     8.6
The Elephant Man                         8.2
                                        ... 
Forrest Gump                             8.7
Le salaire de la peur                    8.1
3 Idiots                                 8.2
Network                                  8.1
Eternal Sunshine of the Spotless Mind    8.3
Name: Rating, Length: 250, dtype: float64

Notice how not only the movie ratings have been returned, but also the name of the movie! This is precisely because we have set the movie title to be the index! For example, if we had asked for the `"Rating"` column of the original DataFrame, `imdb`, we would see:

In [79]:
imdb.get('Rating')

0      8.4
1      8.3
2      8.3
3      8.6
4      8.2
      ... 
245    8.7
246    8.1
247    8.2
248    8.1
249    8.3
Name: Rating, Length: 250, dtype: float64

This is one way in which indices are very useful - they provide meaningful labels for the data.

At first glance, it might look like asking for a column using `.get` returns a DataFrame with one column, but that's not quite right. Instead, it returns a special type of thing called a *Series*:

In [80]:
type(imdb_by_name.get('Rating'))

pandas.Series

You can think of a `Series` as an array with an index. Whereas arrays are simple sequences of numbers without labels, `Series` can have labels. This is often very useful.

`ratings` is now a `Series` which contains the column of movie ratings. Suppose we're interested in the rating of a particular movie: _Alien_. To do so, we will use the `.loc` *accessor* which pulls a value from the Series at a particular *loc*ation:

In [81]:
ratings.loc['Alien']

8.5

There are a couple of things to note here. First, those are square brackets around `"Alien"`. This is because `.loc` is not a method, but an *accessor*. The square brackets signal that we're going to be extracting an element from the `Series`. Second, we passed in the label as a string.

**Question 2.4.1.** Find the rating of _3 Idiots_.

In [82]:
three_idiots_rating = ratings.loc['3 Idiots'] # SOLUTION
three_idiots_rating

8.2

In [83]:
bool(three_idiots_rating == 8.2)

True

Now suppose we wanted to know the year in which _Alien_ was released. We could do this by first getting the column of years:

In [84]:
years = imdb_by_name.get('Year')
years

Title
M                                        1931
Singin' in the Rain                      1952
All About Eve                            1950
Léon                                     1994
The Elephant Man                         1980
                                         ... 
Forrest Gump                             1994
Le salaire de la peur                    1953
3 Idiots                                 2009
Network                                  1976
Eternal Sunshine of the Spotless Mind    2004
Name: Year, Length: 250, dtype: int64

And then using `.loc` to get the right entry:

In [85]:
years.loc['Alien']

1979

We could also do this in one step by *chaining* the operations together:

In [86]:
imdb_by_name.get('Year').loc['Alien']

1979

This works because Python first evaluates `imdb_by_name.get('Year')` to a Series. It then evaluates the `.loc['Alien']` to return the year.

Chaining is used pretty frequently and can be handy. Just be sure not to chain *too* many things together that your code gets hard to read. You can always save an intermediate result to a variable.

**Question 2.4.2** Find the decade in which _Gone Girl_ was released using chaining. 

**_Hint:_**  `imdb_by_name` has a column named `'Decade'`.

In [87]:
decade = imdb_by_name.get('Decade').loc['Gone Girl'] # SOLUTION
decade

2010

In [88]:
bool(decade == 2010)

True

# 3. Analyzing datasets

With just a few DataFrame methods, we can answer some interesting questions about the IMDb dataset.

If we want just the ratings of the movies, we can use `.get`:

In [89]:
ratings = imdb_by_name.get('Rating')
ratings

Title
M                                        8.4
Singin' in the Rain                      8.3
All About Eve                            8.3
Léon                                     8.6
The Elephant Man                         8.2
                                        ... 
Forrest Gump                             8.7
Le salaire de la peur                    8.1
3 Idiots                                 8.2
Network                                  8.1
Eternal Sunshine of the Spotless Mind    8.3
Name: Rating, Length: 250, dtype: float64

Remember that `ratings` is a Series. Series objects have some useful methods.

**Question 3.1.** Find the rating of the highest-rated movie in the dataset.

**_Hint:_**  Type `ratings.` and hit Tab to see a list of the available methods. Is there one that looks useful?

In [90]:
highest_rating = ratings.max() # SOLUTION
highest_rating

9.2

In [91]:
bool(highest_rating == 9.2)

True

You probably want to know the *name* of the movie whose rating you found!  To do that, we can sort the whole Series using the `.sort_values` method:

In [92]:
ratings.sort_values()

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

So there are actually 2 highest-rated movies in the dataset: *The Shawshank Redemption* and *The Godfather*.

Notice that we are sorting by the ratings, not the labels! The label moves with the rating as it is sorted. This is exactly what we want.

When we use the `sort_values` method, the resulting Series has the data sorted in ascending order, from small to large. This is the default behavior of `sort_values`, but we can change that. Had we wanted the highest rated movies on top, we would need to specify that the sorting should not be in ascending order with an optional *keyword argument*:


In [93]:
ratings.sort_values(ascending=False)

Title
The Godfather                             9.2
The Shawshank Redemption                  9.2
The Godfather: Part II                    9.0
12 Angry Men                              8.9
Il buono, il brutto, il cattivo (1966)    8.9
                                         ... 
Monsters, Inc. (2001)                     8.0
The Big Sleep                             8.0
X-Men: Days of Future Past                8.0
Roman Holiday                             8.0
Kumonosu-jô                               8.0
Name: Rating, Length: 250, dtype: float64

If we set the keyword argument `ascending` to `True`, we get the same result as if we did not set it at all. This is what we mean when we say that the default behavior of `sort_values` is to sort in ascending order. Confirm that the next two cells give the same output.

In [94]:
ratings.sort_values(ascending=True)

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

In [95]:
ratings.sort_values()

Title
Akira                               8.0
Per un pugno di dollari             8.0
Guardians of the Galaxy             8.0
The Man Who Shot Liberty Valance    8.0
Underground                         8.0
                                   ... 
Schindler's List                    8.9
12 Angry Men                        8.9
The Godfather: Part II              9.0
The Shawshank Redemption            9.2
The Godfather                       9.2
Name: Rating, Length: 250, dtype: float64

Not only can we sort Series, but we can sort entire DataFrames, too. When we do that, we have to specify the column to sort by:

In [96]:
imdb_by_name.sort_values('Rating')

,Votes,Rating,Year,Decade
Title,,,,
Akira,91652,8.0,1988,1980
Per un pugno di dollari,124671,8.0,1964,1960
Guardians of the Galaxy,527349,8.0,2014,2010
The Man Who Shot Liberty Valance,49135,8.0,1962,1960
Underground,39447,8.0,1995,1990
...,...,...,...,...
Schindler's List,761224,8.9,1993,1990
12 Angry Men,384187,8.9,1957,1950
The Godfather: Part II,692753,9.0,1974,1970


Similarly, we can specify that the sort should be in descending order:

In [97]:
imdb_by_name.sort_values('Rating', ascending=False)

,Votes,Rating,Year,Decade
Title,,,,
The Godfather,1027398,9.2,1972,1970
The Shawshank Redemption,1498733,9.2,1994,1990
The Godfather: Part II,692753,9.0,1974,1970
12 Angry Men,384187,8.9,1957,1950
"Il buono, il brutto, il cattivo (1966)",447875,8.9,1966,1960
...,...,...,...,...
"Monsters, Inc. (2001)",500576,8.0,2001,2000
The Big Sleep,59578,8.0,1946,1940
X-Men: Days of Future Past,427099,8.0,2014,2010



Some details about sorting a DataFrame:

1. The first argument to `sort_values` is the name of a column to sort by.
2. If the column has strings in it, `sort` will sort alphabetically; if the column has numbers, it will sort numerically.
3. `imdb_by_name.sort_values('Rating')` returns a new DataFrame; the `imdb_by_name` DataFrame doesn't get modified. For example, if we called `imdb_by_name.sort('Rating')`, then running `imdb_by_name` by itself would still return the unsorted DataFrame. To save the result, you should assign it to a new variable.
4. Rows always stick together when a DataFrame is sorted.  It wouldn't make sense to sort just one column and leave the other columns alone.  For example, in this case, if we sorted just the `'Rating'` column, the movies would all end up with the wrong ratings.

**Question 3.2.** Create a version of `imdb_by_name` that's sorted chronologically, with the earliest movies first.  Call it `imdb_sorted`.

In [98]:
imdb_sorted = imdb_by_name.sort_values(['Year', 'Votes']) # SOLUTION
imdb_sorted

,Votes,Rating,Year,Decade
Title,,,,
The Kid,55784,8.3,1921,1920
The Gold Rush,58506,8.2,1925,1920
The General,46332,8.2,1926,1920
Metropolis,98794,8.3,1927,1920
M,88355,8.4,1931,1930
...,...,...,...,...
Gone Girl,441174,8.1,2014,2010
Guardians of the Galaxy,527349,8.0,2014,2010
Interstellar,689541,8.6,2014,2010


In [99]:
imdb_sorted.index[1] == 'The Gold Rush'

True

**Question 3.3.** What's the title of the earliest (oldest) movie in the dataset?  You could just look this up from the output of the previous cell.  Instead, write Python code to grab the title from `imdb_sorted` and store it in `earliest_movie_title`.

**_Hint:_**  Remember that the index is an array.

In [100]:
earliest_movie_title = imdb_sorted.index[0] # SOLUTION
earliest_movie_title

'The Kid'

In [101]:
earliest_movie_title == 'The Kid'

True

Suppose we want to get the rating of the oldest movie in the DataFrame. One way to do this is to first find the index label of the oldest movie (which we've already done). We then extract the `"Rating"` column and use `.loc` to find the rating of the oldest movie.

In [102]:
imdb_sorted.get('Rating').loc[earliest_movie_title]

8.3

There's a faster way, though. A Series not only has a `.loc` accessor, but also an `.iloc` accessor. While `.loc` looks up things by *label*, `.iloc` looks up elements by *integer position*.

Let's remember what is in the `'Rating'` column:

In [103]:
imdb_sorted.get('Rating')

Title
The Kid                    8.3
The Gold Rush              8.2
The General                8.2
Metropolis                 8.3
M                          8.4
                          ... 
Gone Girl                  8.1
Guardians of the Galaxy    8.0
Interstellar               8.6
Inside Out (2015/I)        8.5
Mad Max: Fury Road         8.3
Name: Rating, Length: 250, dtype: float64

If we want the rating of the first row, we can use `.iloc[0]`:

In [104]:
imdb_sorted.get('Rating').iloc[0]

8.3

This returns the exact same thing as `imdb_sorted.get('Rating').loc['The Kid']`; these are two ways of doing the same thing. Usually it is more convenient to access an element by its label rather than by its integer position, but both `.loc` and `.iloc` are good to know.

**Question 3.4.** What is the rating of the fifth oldest movie in the dataset? You could just look this up from the output of the previous cell. Instead, write Python code to find out.

In [105]:
fifth_oldest_rating = imdb_sorted.get('Rating').iloc[4] # SOLUTION
fifth_oldest_rating

8.4

In [106]:
bool(fifth_oldest_rating == 8.4 or fifth_oldest_rating == 8.5)

True

# 4. Finding pieces of a dataset

Suppose you're interested in movies from the 1950s.  Sorting the DataFrame by year doesn't help you, because the 1950s are in the middle of the dataset. Instead, we'll use a feature of Series that allows us to easily compare each element in a column to a particular value.

First, remember that we can use `.get` to extract a single column. The result is not a DataFrame, but rather a Series:

In [107]:
imdb_by_name.get('Decade')

Title
M                                        1930
Singin' in the Rain                      1950
All About Eve                            1950
Léon                                     1990
The Elephant Man                         1980
                                         ... 
Forrest Gump                             1990
Le salaire de la peur                    1950
3 Idiots                                 2000
Network                                  1970
Eternal Sunshine of the Spotless Mind    2000
Name: Decade, Length: 250, dtype: int64

We want to check whether each movie is released in the decade 1950. Python gives us a way of checking whether two things are equal with `==` (remember that `=` is already being used for another purpose: it assigns values to variable names):

In [108]:
3 == 4

False

In [109]:
3 == 3

True

`True` and `False` are instances of a type that we haven't seen before:

In [110]:
type(True)

bool

`bool` stands for "Boolean", named after the English logician [George Boole](https://en.wikipedia.org/wiki/George_Boole). We say that "True" and "False" are *Boolean* values.

It turns out that we can easily check if *each* of the elements in a `Series` is equal to something:

In [111]:
imdb_by_name.get('Decade') == 1950

Title
M                                        False
Singin' in the Rain                       True
All About Eve                             True
Léon                                     False
The Elephant Man                         False
                                         ...  
Forrest Gump                             False
Le salaire de la peur                     True
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

We see that the result is a new series which has `True` only where the decade was 1950, and `False` everywhere else. We say that the resulting series is a series of *Booleans*, or a *Boolean Series*.

Let's call this result `is_from_1950s`. Its name can be read like it is a question: "is this movie from the 1950s"?

In [112]:
is_from_1950s = imdb_by_name.get('Decade') == 1950
is_from_1950s

Title
M                                        False
Singin' in the Rain                       True
All About Eve                             True
Léon                                     False
The Elephant Man                         False
                                         ...  
Forrest Gump                             False
Le salaire de la peur                     True
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

Each row is an answer to this question. Is _The Elephant Man_ from the 1950s? `False`. Is _All About Eve_ from the 1950s? `True`.

We can use `is_from_1950s` to select only the rows from `imdb_by_name` for which the answer is `True`. The syntax for this is:

In [113]:
imdb_by_name[is_from_1950s]

,Votes,Rating,Year,Decade
Title,,,,
Singin' in the Rain,132823,8.3,1952,1950
All About Eve,74178,8.3,1950,1950
Some Like It Hot,156432,8.3,1959,1950
The Killing,56671,8.0,1956,1950
Roman Holiday,87437,8.0,1953,1950
Touch of Evil,65408,8.1,1958,1950
Rashômon,90434,8.3,1950,1950
La strada,42446,8.0,1954,1950
North by Northwest,198795,8.4,1959,1950


What `imdb_by_name[is_from_1950s]` does, precisely, is to go through `imdb_by_name` row by row. If the row named _Singin' in the Rain_ has the value `True` in `is_from_1950s`, that row is kept. If the value is `False`, the row is discarded. And so on, for every row.

Note that we could have accomplished this without ever creating the variable `is_from_1950s` by simply placing the code that we used to create the boolean series directly inside the `[...]`. This is a typical pattern you'll be using a lot!

In [114]:
imdb_by_name[imdb_by_name.get('Decade') == 1950]

,Votes,Rating,Year,Decade
Title,,,,
Singin' in the Rain,132823,8.3,1952,1950
All About Eve,74178,8.3,1950,1950
Some Like It Hot,156432,8.3,1959,1950
The Killing,56671,8.0,1956,1950
Roman Holiday,87437,8.0,1953,1950
Touch of Evil,65408,8.1,1958,1950
Rashômon,90434,8.3,1950,1950
La strada,42446,8.0,1954,1950
North by Northwest,198795,8.4,1959,1950


It helps to read the square brackets as "where." So the command in the cell above says to keep all rows from `imbdb_by_name` *where* the decade is the 1950s. 

Creating a new DataFrame by selecting only certain rows from an existing DataFrame which satisfy some condition is called *querying*. The line of code `imdb_by_name[imdb_by_name.get('Decade') == 1950]` is a *query*.

**Question 4.1.** Create a DataFrame called `ninety_eight` containing the movies that came out in 1998.

In [115]:
ninety_eight = imdb_by_name[imdb_by_name.get('Year') == 1998] # SOLUTION
ninety_eight

,Votes,Rating,Year,Decade
Title,,,,
Saving Private Ryan,769893,8.5,1998,1990
American History X,694602,8.5,1998,1990
"Lock, Stock and Two Smoking Barrels (1998)",372863,8.2,1998,1990
The Big Lebowski,473988,8.2,1998,1990
The Truman Show,583004,8.0,1998,1990


In [116]:
ninety_eight.index[0] == 'Saving Private Ryan'

True

So far we've only been finding where a column is *exactly* equal to a certain value. However, there are many other comparison operators we could use.  Here are a few:

|Operator|Tests|
|-|-|
|`==`|thing on left is equal to thing on right|
|`!=`|thing on left is *not* equal to thing on right|
|`>`|thing on left is greater than (and not equal to) thing on right|
|`>=`|thing on left is greater than or equal to thing on right|
|`<`|thing on left is less than (and not equal to) thing on right|
|`<=`|thing on left is less than or equal to thing on right|


**Question 4.2.** Using operators from the table above, find all the movies with a rating higher than 8.6.  Put their data in a DataFrame called `really_highly_rated`.

In [117]:
really_highly_rated = imdb_by_name[imdb_by_name.get('Rating') > 8.6] # SOLUTION
really_highly_rated

,Votes,Rating,Year,Decade
Title,,,,
The Godfather,1027398,9.2,1972,1970
The Shawshank Redemption,1498733,9.2,1994,1990
"Il buono, il brutto, il cattivo (1966)",447875,8.9,1966,1960
The Lord of the Rings: The Two Towers,967389,8.7,2002,2000
The Dark Knight,1473049,8.9,2008,2000
Inception,1271949,8.7,2010,2010
Fight Club,1177098,8.8,1999,1990
The Godfather: Part II,692753,9.0,1974,1970
Shichinin no samurai,206216,8.7,1954,1950


In [118]:
really_highly_rated.shape == (20, 4)

True

What is the highest rating of any movie from the 1990s? We now have the tools to answer questions like these. Breaking it into pieces, we first find all of the movies from the 1990s:

In [119]:
is_from_1990s = imdb_by_name.get('Decade') == 1990
is_from_1990s

Title
M                                        False
Singin' in the Rain                      False
All About Eve                            False
Léon                                      True
The Elephant Man                         False
                                         ...  
Forrest Gump                              True
Le salaire de la peur                    False
3 Idiots                                 False
Network                                  False
Eternal Sunshine of the Spotless Mind    False
Name: Decade, Length: 250, dtype: bool

We then select only these movies from our DataFrame:

In [120]:
from_1990s = imdb_by_name[is_from_1990s]
from_1990s

,Votes,Rating,Year,Decade
Title,,,,
Léon,635139,8.6,1994,1990
Mononoke-hime,192165,8.4,1997,1990
Saving Private Ryan,769893,8.5,1998,1990
In the Name of the Father,95212,8.1,1993,1990
Before Sunrise,158867,8.0,1995,1990
The Silence of the Lambs,767224,8.6,1991,1990
The Shawshank Redemption,1498733,9.2,1994,1990
Heat,388239,8.2,1995,1990
Toy Story,535249,8.3,1995,1990


We then find the highest rating out of just these movies:

In [121]:
from_1990s.get('Rating').max()

9.2

Or, if we wanted to do all of this more concisely using chaining:

In [122]:
imdb_by_name[imdb_by_name.get('Decade') == 1990].get('Rating').max()

9.2

**Question 4.3.** Find the average rating for movies released in the 20th century and the average rating for movies released in the 21st century for the movies in `imdb`.

**_Hint:_**  Series have a `.mean()` method. Note that the year 2000 is the last year included in the 20th century, and that the earliest movie in the dataset is from 1921!

In [123]:
average_20th_century_rating = imdb_by_name[imdb_by_name.get('Year') <= 2000].get('Rating').mean() # SOLUTION
average_20th_century_rating

8.280113636363636

In [124]:
bool(np.isclose(average_20th_century_rating, 8.280113636363636))

True

In [125]:
average_21st_century_rating = imdb_by_name[imdb_by_name.get('Year') > 2000].get('Rating').mean() # SOLUTION
average_21st_century_rating

8.23108108108108

In [126]:
bool(np.isclose(average_21st_century_rating, 8.23108108108108))

True

The attribute `shape` tells you how many rows and columns are in a DataFrame. (An attribute, or a property, is not a method and is not called using parentheses.)

In [127]:
imdb_by_name.shape

(250, 4)

Like an array, you can get the first element of the shape using `[0]`, and the second element using `[1]`. For instance, the number of rows in `imdb_by_name` is:

In [128]:
imdb_by_name.shape[0]

250

We can use this to answer "How many movies are from the 20th century?":

In [129]:
imdb_by_name[imdb_by_name.get('Year') <= 2000].shape[0]

176

**Question 4.4.** Use `shape` (and arithmetic) to find the *proportion* of movies in the dataset from the 20th century, and the proportion from the 21st century.

In [130]:
proportion_in_20th_century = imdb_by_name[imdb_by_name.get('Year') <= 2000].shape[0] / imdb_by_name.shape[0] # SOLUTION
proportion_in_20th_century

0.704

In [131]:
bool(np.isclose(proportion_in_20th_century, 0.704))

True

In [132]:
proportion_in_21st_century = imdb_by_name[imdb_by_name.get('Year') > 2000].shape[0] / imdb_by_name.shape[0] # SOLUTION
proportion_in_21st_century

0.296

In [133]:
bool(np.isclose(proportion_in_21st_century, 0.296))

True

**Question 4.5.** Finally, let's revisit the `population_by_year` DataFrame from earlier in the lab.  Compute the year when the world population first went above 7 billion.

In [134]:
year_population_crossed_7_billion = population_by_year[population_by_year.get('Population') > 7_000_000_000].index[0] # SOLUTION
year_population_crossed_7_billion

2011

In [135]:
bool(year_population_crossed_7_billion == 2011)

True

# Finish Line 🏁

Congratulations! You are done with Lab 1.

In [136]:
# For your convenience, you can run this cell to run all the tests at once!
grader.check_all()